### PR TITLE
Verify upgrades change node/master software version

### LINKED ISSUE #7914: test: ensure node upgrades between k8s versions work
This is blocked by using only MIG templates #7912 and a mechanism to do updates #6088.

Overview:

- this needs a separate Jenkins job to run continuously on a different cluster (shouldn't be part of e2es yet because of version skewed testing)
- as a minimum bar, this should test only from the latest released version to head
- as a minimum bar, this should run e2es **after** the upgrade happens

### PR DESCRIPTION
**Pull request by mbforbes**
Fixes #7914 and #8081.

Risk assessment for this PR: **low**. It adds more rigorous checking to an e2e test.

I'll hope to get my two LGTMs from:
- @roberthbailey (reviewed previous PRs for this file)
- @davidopp (TL (I think) for @mikedanese who is working on this area too)

### CODE REVIEW

File: test/e2e/cluster_upgrade.go [hunk 1]
```diff
@@ -369,20 +374,56 @@ func testMasterUpgrade(ip, v string, mUp func(v string) error) {
...
+	}
+	// We do prefix trimming and then matching because:
+	// want looks like:  0.19.3-815-g50e67d4
+	// got  looks like: v0.19.3-815-g50e67d4034e858-dirty
+	got := strings.TrimPrefix(v.GitVersion, "v")
+	if !strings.HasPrefix(got, want) {
+		return fmt.Errorf("master had kube-apiserver version %s which does not start with %s",
+			got, want)
+	}
+	Logf("Master is at version %s", want)
```
Comment (davidopp) [inline test/e2e/cluster_upgrade.go]: should this be got?

File: test/e2e/cluster_upgrade.go [hunk 2]
```diff
@@ -262,6 +264,8 @@ var _ = Describe("Skipped", func() {
 				expectNoError(validate(f, svcName, rcName, ingress, replicas))
 				By("Performing a master upgrade")
 				testMasterUpgrade(ip, v, masterUpgrade)
+				By("Checking master version")
```
Comment (davidopp) [inline test/e2e/cluster_upgrade.go]: should you also add this check for the kube-push case?

File: test/e2e/cluster_upgrade.go [hunk 3]
```diff
@@ -369,20 +374,56 @@ func testMasterUpgrade(ip, v string, mUp func(v string) error) {
...
 	expectNoError(nUp(f, replicas, v))
 	Logf("Node upgrade complete")
+	By("Checking node versions")
+	expectNoError(checkNodeVersions(f.Client, v))
+	Logf("All nodes are at version %s", v)
+}
 
-	// TODO(mbforbes): Validate that:
-	// - the node software version truly changed
+func checkNodeVersions(c *client.Client, want string) error {
```
Comment (davidopp) [inline test/e2e/cluster_upgrade.go]: rename to checkNodesVersions ?

File: test/e2e/cluster_upgrade.go [hunk 4]
```diff
@@ -369,20 +374,56 @@ func testMasterUpgrade(ip, v string, mUp func(v string) error) {
...
+		// kv/kvp look  like: v0.19.3-815-g50e67d4034e858-dirty
+		kv, kpv := strings.TrimPrefix(n.Status.NodeInfo.KubeletVersion, "v"),
+			strings.TrimPrefix(n.Status.NodeInfo.KubeProxyVersion, "v")
+		if !strings.HasPrefix(kv, want) {
+			return fmt.Errorf("node %s had kubelet version %s which does not start with %s",
+				n.ObjectMeta.Name, kv, want)
+		}
+		if !strings.HasPrefix(kpv, want) {
+			return fmt.Errorf("node %s had kube-proxy version %s which does not start with %s",
+				n.ObjectMeta.Name, kv, want)
```
Comment (davidopp) [inline test/e2e/cluster_upgrade.go]: kpv

**Comment by roberthbailey** [comment]
LGTM after you address @davidopp's comments.

File: test/e2e/cluster_upgrade.go (hunk 1 above)
Comment (mbforbes) [inline test/e2e/cluster_upgrade.go]: Good question. I'm trying to be consistent about logging the shorter version. (They are really the same thing.)

File: test/e2e/cluster_upgrade.go (hunk 2 above)
Comment (mbforbes) [inline test/e2e/cluster_upgrade.go]: Another good question. Kube-push uses local binaries, so I don't think the whole versioning situation is as well defined as passing an explicit version (as in all other non-kube-push cases). We might be able to shell out and check the latest git commit in the user's repo (assuming this test is running in a git repo), but I opted to just address only "official" upgrade paths in this PR. Is this OK?

File: test/e2e/cluster_upgrade.go (hunk 3 above)
Comment (mbforbes) [inline test/e2e/cluster_upgrade.go]: Double plural reads weird to me but I guess makes more sense. Done (pushing soon).

File: test/e2e/cluster_upgrade.go (hunk 4 above)
Comment (mbforbes) [inline test/e2e/cluster_upgrade.go]: Great catch. Done (pushing soon).

**Comment by mbforbes** [comment]
Comments addressed—thanks guys. PTAL.

**Comment by davidopp** [comment]
LGTM

**Comment by mbforbes** [comment]
Well, seeing as I'm now oncall... :-)

**Pull request merged**